usdigitalresponse/usdr-gost
# Feat: Use `window.sessionStorage` to override feature flags

## Description

This PR adds the ability to store persistent overrides to a feature flag's for the current page session (tab), which is useful for overriding/toggling feature flags in non-development environments in scenarios where the tab may be reloaded after the override is in place. Previously, feature flags could only be overridden by directly assigning a value, e.g. to `window.APP_CONFIG.featureFlags.useNewTable = false`, which does not survive page reloads. This PR provides a new function, `window.APP_CONFIG.overrideFeatureFlag()`, which uses [`window.sessionStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) to persist overrides within the current tab.

Note that the `window.APP_CONFIG.featureFlags` may still be modified directly during development (which is fine as long as you don't reload the page).

Summary of changes:
- Update the `getFeatureFlags()` utility function in `packages/client/src/helpers/featureFlags/utils.js` to check for and deserialize any feature flag overrides persisted to `window.sessionStorage` under the `featureFlags` storage key.
- Add a new `window.APP_CONFIG.overrideFeatureFlag()` helper function in the dev-only `packages/client/public/deploy-config.js` file, as well as in the Terraform template file (at `terraform/modules/gost_website/tpl/deploy-config.js`), which sets, serializes, and persists overrides to `window.sessionStorage` under the `featureFlags` storage key.
- Updated unit tests in `packages/client/tests/unit/helpers/featureFlags.spec.js` to account for the possibility of overrides.
- Updated documentation at `docs/feature-flags.md` with a new "Dynamically Modifying Flag Values" section that provides instructions for setting and clearing feature flag overrides.

## Screenshots / Demo Video

https://github.com/usdigitalresponse/usdr-gost/assets/1851017/81357a6e-5a43-4e4b-94d1-7f9d9a4b1e30

## Testing

1. In a local development environment, ensure that the `useNewTable` feature flag is set to `true` in `packages/client/public/deploy-config.js`, then start the local client and server.
2. Navigate to the app, and log in if necessary. Verify that the "Eligibility Codes" and "Keywords" tab items in the nav menu are not present.
3. Navigate to the "Browse Grants" page in the app and verify that the view is using the new table that provides the redesigned search experience.
4. Open the browser's development console, and run the following:
    ```console
    > window.APP_CONFIG.overrideFeatureFlag('useNewTable', false)
    // Log output: New feature flag overrides in page session: {"useNewTable":false}
    ```
5. Reload the current tab and verify that the "Eligibility Codes" and "Keywords" menu items are visible, and that the "Browse Grants" page is not using the table that provides the redesigned search experience.
6. In a separate tab (do not use the "Duplicate Tab" feature in your browser, as this will copy the `window.sessionStorage` object), navigate verify that the nav menu items and "Browse Grants" page appear as they did in Step 2 (i.e. using the redesigned experience, indicating that the `useNewTable` feature flag is active).

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
